File: frontend/src/api/batches.ts
// Batches API hooks. Backed by /api/v2/batches.
//
// One read-side endpoint today: list-all-for-store with sort
// + precomputed transfers_total + variance + transfer_count.
// Write-side (create / edit / link transfers) stays on the
// legacy Flask path.

import { useQuery } from "@tanstack/react-query";

import { api } from "../lib/api";
import { getCurrentIdentity } from "../lib/auth";

export interface BatchRow {
  id: number;
  ach_date: string;
  company: string;
  batch_ref: string;
  ach_amount: number;
  status: string;
  reconciled: boolean;
  transfer_dates: string;
  notes: string;
  transfers_total: number;
  variance: number;
  transfer_count: number;
}

export interface BatchListResponse {
  rows: BatchRow[];
}

export type BatchSort =
  | "ach_date" | "company" | "batch_ref" | "ach_amount" | "status" | "";
export type BatchDir = "asc" | "desc";

export interface BatchWriteBody {
  ach_date: string;
  company: string;
  batch_ref: string;
  ach_amount: number;
  transfer_dates?: string;
  status?: string;
  reconciled?: boolean;
  notes?: string;
}

export function useBatch(batchId: number | undefined) {
  const identity = getCurrentIdentity();
  return useQuery<{ batch: BatchRow }>({
    enabled:
      Number.isFinite(batchId) &&
      identity?.store_id != null,
    queryKey: ["batches", "detail", identity?.store_id, batchId],
    queryFn: () => api<{ batch: BatchRow }>(`/api/v2/batches/${batchId}`),
  });
}

export async function createBatch(
  body: BatchWriteBody,
): Promise<{ batch: BatchRow }> {
  return api<{ batch: BatchRow }>(
    "/api/v2/batches",
    { method: "POST", json: body },
  );
}

export async function updateBatch(
  batchId: number, body: BatchWriteBody,
): Promise<{ batch: BatchRow }> {
  return api<{ batch: BatchRow }>(
    `/api/v2/batches/${batchId}`,
    { method: "PUT", json: body },
  );
}

export interface BatchLinkedTransferRow {
  id: number;
  send_date: string;
  company: string;
  sender_name: string;
  recipient_name: string;
  country: string;
  confirm_number: string;
  send_amount: number;
  fee: number;
  federal_tax: number;
  total_collected: number;
  status: string;
}

export function useBatchTransfers(batchId: number | undefined) {
  const identity = getCurrentIdentity();
  return useQuery<{ transfers: BatchLinkedTransferRow[] }>({
    enabled:
      Number.isFinite(batchId) &&
      identity?.store_id != null,
    queryKey: ["batches", "transfers", identity?.store_id, batchId],
    queryFn: () =>
      api<{ transfers: BatchLinkedTransferRow[] }>(
        `/api/v2/batches/${batchId}/transfers`,
      ),
  });
}

export function useBatches(sort: BatchSort = "", direction: BatchDir = "desc") {
  const identity = getCurrentIdentity();
  const enabled = identity?.store_id != null;
  return useQuery<BatchListResponse>({
    enabled,
    queryKey: ["batches", "list", identity?.store_id, sort, direction],
    queryFn: () => {
      const params = new URLSearchParams();
      if (sort) params.set("sort", sort);
      params.set("direction", direction);
      return api<BatchListResponse>(
        `/api/v2/batches?${params.toString()}`,
      );
    },
  });
}

File: api/Modules/Batches/Requests/batches.py
"""Pydantic schemas for the ACH batches read-side."""
from pydantic import BaseModel, ConfigDict, Field


class BatchRow(BaseModel):
    """One ACH batch row. Mirrors the columns the legacy
    `/batches` table renders + the computed totals
    (transfers_total, variance, transfer_count) the SPA needs
    to colour the variance cell."""

    model_config = ConfigDict(extra="forbid")

    id: int
    ach_date: str         # YYYY-MM-DD
    company: str
    batch_ref: str
    ach_amount: float
    status: str = "Pending"
    reconciled: bool = False
    transfer_dates: str = ""
    notes: str = ""
    # Server-precomputed totals so the SPA doesn't have to N+1
    # the transfers table per row.
    transfers_total: float
    variance: float
    transfer_count: int


class BatchListResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    rows: list[BatchRow] = Field(default_factory=list)


class BatchResponse(BaseModel):
    """Single-batch payload used by the create / edit / detail
    endpoints. Same shape as a row from the list endpoint."""

    model_config = ConfigDict(extra="forbid")

    batch: BatchRow


class BatchWriteRequest(BaseModel):
    """POST/PUT body for /batches. Mirrors the legacy form's
    fields. Server pads `transfer_count` / `variance` /
    `transfers_total` from the live Transfer table after the
    write."""

    model_config = ConfigDict(extra="forbid")

    ach_date:        str = Field(..., min_length=10, max_length=10)  # YYYY-MM-DD
    company:         str = Field(..., min_length=1, max_length=30)
    batch_ref:       str = Field(..., min_length=1, max_length=60)
    ach_amount:      float = Field(..., ge=0)
    transfer_dates:  str = Field("", max_length=60)
    status:          str = Field("Pending", max_length=30)
    reconciled:      bool = False
    notes:           str = Field("", max_length=255)


class BatchTransfersResponse(BaseModel):
    """Transfers linked to one ACH batch by `batch_id ==
    batch_ref`. Same TransferRow shape the Transfers module
    uses, mirrored here so the Batches read-side stays
    self-contained."""

    model_config = ConfigDict(extra="forbid")

    transfers: list["BatchLinkedTransferRow"]


class BatchLinkedTransferRow(BaseModel):
    """Subset of Transfer columns the SPA's batch-detail page
    renders. Kept narrower than the full TransferRow because
    cross-module Pydantic re-export gets messy and the SPA
    doesn't need every field here."""

    model_config = ConfigDict(extra="forbid")

    id: int
    send_date: str
    company: str
    sender_name: str
    recipient_name: str = ""
    country: str = ""
    confirm_number: str = ""
    send_amount: float
    fee: float = 0.0
    federal_tax: float = 0.0
    total_collected: float
    status: str = "Sent"


# Forward reference resolution
BatchTransfersResponse.model_rebuild()
